mypy/checker.py

```diff
@@ -407,49 +407,63 @@ def _visit_overloaded_func_def(self, defn: OverloadedFuncDef) -> None:
+                    assert isinstance(item2, Decorator)
+                    sig2 = self.function_type(item2.func)
+
+                    assert isinstance(sig1, CallableType)
+                    assert isinstance(sig2, CallableType)
+
+                    if not are_argument_counts_overlapping(sig1, sig2):
+                        continue
 
-                assert isinstance(impl_type, CallableType)
-                assert isinstance(sig1, CallableType)
-                if not is_callable_subtype(impl_type, sig1, ignore_return=True):
-                    self.msg.overloaded_signatures_arg_specific(i + 1, defn.impl)
-                impl_type_subst = impl_type
-                if impl_type.variables:
-                    unified = unify_generic_callable(impl_type, sig1, ignore_return=False)
-                    if unified is None:
-                        self.fail("Type variable mismatch between " +
-                                  "overload signature {} and implementation".format(i + 1),
-                                  defn.impl)
+                    if overload_can_never_match(sig1, sig2):
+                        self.msg.overloaded_signature_will_never_match(
+                            i + 1, i + j + 2, item2.func)
+                    elif is_unsafe_overlapping_overload_signatures(sig1, sig2):
+                        self.msg.overloaded_signatures_overlap(
+                            i + 1, i + j + 2, item.func)
+                if defn.impl:
+                    if isinstance(defn.impl, FuncDef):
+                        impl_type = defn.impl.type
+                    elif isinstance(defn.impl, Decorator):
+                        impl_type = defn.impl.var.type
+                    else:
+                        assert False, "Impl isn't the right type"
+                    # This can happen if we've got an overload with a different
+                    # decorator too -- we gave up on the types.
+                    if impl_type is None or isinstance(impl_type, AnyType) or sig1 is None:
                         return
-                    impl_type_subst = unified
-                if not is_subtype(sig1.ret_type, impl_type_subst.ret_type):
-                    self.msg.overloaded_signatures_ret_specific(i + 1, defn.impl)
+
+                    assert isinstance(impl_type, CallableType)
+                    assert isinstance(sig1, CallableType)
+                    if not is_callable_compatible(impl_type, sig1,
+                                                  is_compat=is_subtype, ignore_return=True):
+                        self.msg.overloaded_signatures_arg_specific(i + 1, defn.impl)
+                    impl_type_subst = impl_type
+                    if impl_type.variables:
+                        unified = unify_generic_callable(impl_type, sig1, ignore_return=False)
+                        if unified is None:
+                            self.fail("Type variable mismatch between " +
```

Comment:
Yeah, I don't really understand the logic here either. TBH, I didn't look too hard at it: the only changes I made in my first pass was to just indent everything by a level (because of the strict-optional with statement).

I decided to just remove this error message in my second pass: I was able to find only one test case that produced this error message. (And that test case looked ill-defined anyways -- it was checking the `f(x: Any) -> T` case I was asking about on gitter earlier.)

(Same sort of thing for your comment about the three conditions up above: that was old logic I decided to just keep. I did remove the `sig1 is None` bit though -- I think that condition is actually always going to be false.)